llama_index/schema.py

```diff
@@ -119,3 +147,303 @@ def extra_info_str(self) -> Optional[str]:
+            raise ValueError("object_hash is not set.")
+        return self.object_hash
+
+    @property
+    def object_id(self) -> str:
+        return self._id
+
+    @property
+    def is_id_none(self) -> bool:
+        """Check if doc_id is None."""
+        return self.is_id_none is None
+
+    @property
+    def is_content_none(self) -> bool:
+        """Check if text is None."""
+        return self.content is None
+
+    def get_embedding(self) -> List[float]:
+        """Get embedding.
+
+        Errors if embedding is None.
+
+        """
+        if self.embedding is None:
+            raise ValueError("embedding not set.")
+        return self.embedding
+
+    @property
+    def metadata_str(self) -> str:
+        """Extra info string."""
+        if self.metadata is None:
+            return ""
+
+        return "\n".join([f"{k}: {str(v)}" for k, v in self.metadata.items()])
+
+    @property
+    def source_object_id(self) -> Optional[str]:
+        """Source object id.
+
+        Extracted from the relationships field.
+
+        """
+        return self.relationships.get(DataRelationship.SOURCE, None)
+
+    @property
+    def prev_object_id(self) -> str:
+        """Prev object id."""
+        if DataRelationship.PREVIOUS not in self.relationships:
+            raise ValueError("Object does not have previous link")
+        if not isinstance(self.relationships[DataRelationship.PREVIOUS], str):
+            raise ValueError("Previous object must be a string id")
+        return self.relationships[DataRelationship.PREVIOUS]
+
+    @property
+    def next_object_id(self) -> str:
+        """Next object id."""
+        if DataRelationship.NEXT not in self.relationships:
+            raise ValueError("Object does not have next link")
+        if not isinstance(self.relationships[DataRelationship.NEXT], str):
+            raise ValueError("Next object must be a string id")
+        return self.relationships[DataRelationship.NEXT]
+
+    @property
+    def parent_object_id(self) -> str:
+        """Parent object id."""
+        if DataRelationship.PARENT not in self.relationships:
+            raise ValueError("Object does not have parent link")
+        if not isinstance(self.relationships[DataRelationship.PARENT], str):
+            raise ValueError("Parent object must be a string id")
+        return self.relationships[DataRelationship.PARENT]
+
+    @property
+    def child_object_ids(self) -> List[str]:
+        """Child object ids."""
+        if DataRelationship.CHILD not in self.relationships:
+            raise ValueError("Object does not have child objects")
+        if not isinstance(self.relationships[DataRelationship.CHILD], list):
+            raise ValueError("Child objects must be a list ids")
+        return self.relationships[DataRelationship.CHILD]
+
+    @property
+    def extra_info(self) -> Dict[str, Any]:
+        return self.metadata
+
+
+@dataclass
+class BaseNode(BaseIndexObject):
+    content: str = ""
```

Comment:
out of curiosity, why do this and have base class be more general?

Reply:
I refactored the naming. Everything is now `BaseNode` typed